enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/grpc/ExtAuthService.java

```diff
@@ -96,46 +96,38 @@ public void check(CheckRequest request, StreamObserver<CheckResponse> responseOb
         if (responseObject.isDirectResponse()) {
+            DeniedResponsePreparer deniedResponsePreparer = new DeniedResponsePreparer(DeniedHttpResponse.newBuilder());
             // set headers
             if (responseObject.getHeaderMap() != null) {
                 responseObject.getHeaderMap().forEach((key, value) -> {
                             HeaderValueOption headerValueOption = HeaderValueOption.newBuilder()
                                     .setHeader(HeaderValue.newBuilder().setKey(key).setValue(value).build())
                                     .build();
-                            deniedResponseBuilder.addHeaders(headerValueOption);
+                            deniedResponsePreparer.addHeaders(headerValueOption);
                         }
                 );
             }
-            deniedResponseBuilder.addHeaders(HeaderValueOption.newBuilder().setHeader(HeaderValue.newBuilder()
+            deniedResponsePreparer.addHeaders(HeaderValueOption.newBuilder().setHeader(HeaderValue.newBuilder()
                     .setKey(APIConstants.API_TRACE_KEY).setValue(traceKey).build()));
 
             // set status code
             HttpStatus status = HttpStatus.newBuilder().setCodeValue(responseObject.getStatusCode()).build();
-            deniedResponseBuilder.setStatus(status);
+            deniedResponsePreparer.setStatus(status);
 
             // set body content
             if (responseObject.getResponsePayload() != null) {
-                deniedResponseBuilder.setBody(responseObject.getResponsePayload());
+                deniedResponsePreparer.setBody(responseObject.getResponsePayload());
             }
 
             // set error response body content and headers
             if (responseObject.getErrorCode() != null) {
-                // Error handling create an application/json error payload
-                JSONObject responseJson = new JSONObject();
-                responseJson.put(APIConstants.MessageFormat.ERROR_CODE, responseObject.getErrorCode());
-                responseJson.put(APIConstants.MessageFormat.ERROR_MESSAGE, responseObject.getErrorMessage());
-                responseJson.put(APIConstants.MessageFormat.ERROR_DESCRIPTION, responseObject.getErrorDescription());
-                deniedResponseBuilder.setBody(responseJson.toString());
-
-                HeaderValueOption headerValueOption = HeaderValueOption.newBuilder().setHeader(
-                                HeaderValue.newBuilder().setKey(APIConstants.CONTENT_TYPE_HEADER)
-                                        .setValue(APIConstants.APPLICATION_JSON).build())
-                        .build();
-                deniedResponseBuilder.addHeaders(headerValueOption);
+                // Error handling create an XML(text/xml or application/soap+xml based on SOAP protocol version)
+                // payload if SoapErrorInXMLEnabled is true
+                // otherwise it will create an application/json error payload
```

Comment:
```suggestion
                // payload if SoapErrorInXMLEnabled is true.
                // Otherwise it will create an application/json error payload.
```

Reply:
fixed in https://github.com/wso2/product-microgateway/pull/2972/commits/5c91048f9fa451e4b2f2d91dac704d3f6b755cd3